cluster-api/machines-api/types.go

```diff
@@ -0,0 +1,160 @@
+// GCE) and take the appropriate actions.
+//
+// Any updates to the MachineSpec will be actuated to change the Node in
+// place or replace the Node with one conforming to the spec. In this model,
+// the fact that a controller is able to upgrade a node via in-place upgrades
+// or via a cloud replacement is an implementation detail without API controls.
+//
+// It is recommended, but not required, that provider-specific controllers add
+// finalizers to Machine objects so that they can be triggered on deletion to
+// release the necessary external resources, reporting any errors encountered.
+type Machine struct {
+	metav1.TypeMeta
+	metav1.ObjectMeta
+
+	Spec   MachineSpec
+	Status MachineStatus
+}
+
+type MachineSpec struct {
+	// This ObjectMeta will autopopulate the Node created. Use this to
+	// indicate what labels, annotations, name prefix, etc., should be used
+	// when creating the Node.
+	// +optional
+	metav1.ObjectMeta
+
+	// Provider-specific serialized configuration to use during node
+	// creation. It is recommended that providers maintain their own
+	// versioned API types that should be serialized/deserialized from this
+	// field, akin to component config.
+	// +optional
+	ProviderConfig string
+
+	// A list consisting of "Master" and/or "Node".
+	//
+	//                 +-----------------------+------------------------+
+	//                 | Master present        | Master absent          |
+	// +---------------+-----------------------+------------------------|
+	// | Node present: | Install control plane | Join the cluster as    |
+	// |               | and be schedulable    | just a node            |
+	// |---------------+-----------------------+------------------------|
+	// | Node absent:  | Install control plane | Invalid configuration  |
+	// |               | and be unscheduleable |                        |
+	// +---------------+-----------------------+------------------------+
+	Roles []string
+
+	// Versions of key software to use.
+	// +optional
+	Versions MachineVersionInfo
+
+	// To populate in the associated Node for dynamic kubelet config. This
+	// field already exists in Node, so any updates to it in the Machine
+	// spec will be automatially copied to the linked NodeRef from the
+	// status. The rest of dynamic kubelet config support should then work
+	// as-is.
+	// +optional
+	ConfigSource *corev1.NodeConfigSource
+}
+
+type MachineStatus struct {
+	// If the corresponding Node exists, this will point to its object.
+	// +optional
+	NodeRef *api.ObjectReference
+
+	// When was this status last observed
+	// +optional
+	LastUpdated metav1.Time
+
+	// Indicates whether or not the Machine is fully reconciled. When a
+	// controller observes that the spec has changed and no longer matches
+	// reality, it should update Ready to false before reconciling the
+	// state, and then set back to true when the state matches the spec.
+	Ready bool
```

Comment:
What criteria will govern this value?
Also is there an expectation that this value will always be accurate?

Reply:
This feels like it might be a duplication of data that exists already in API.  Moreover, what is actually updating this?  My take would be this would be a lot of overhead, and may not scale.  Making this generic seems to be an issue.  Can we rely on kubelet ready status, or what other options do we have?